lib/std/fs.zig

```diff
@@ -2035,54 +2076,215 @@ pub const Dir = struct {
     /// this function recursively removes its entries and then tries again.
     /// This operation is not atomic on most file systems.
     pub fn deleteTree(self: Dir, sub_path: []const u8) DeleteTreeError!void {
-        start_over: while (true) {
-            var got_access_denied = false;
+        if (self.deleteFile(sub_path)) {
+            return;
+        } else |err| switch (err) {
+            error.FileNotFound => return,
+            error.IsDir => {},
+            error.AccessDenied,
+            error.InvalidUtf8,
+            error.SymLinkLoop,
+            error.NameTooLong,
+            error.SystemResources,
+            error.ReadOnlyFileSystem,
+            error.NotDir,
+            error.FileSystem,
+            error.FileBusy,
+            error.BadPathName,
+            error.Unexpected,
+            => |e| return e,
+        }
 
-            // First, try deleting the item as a file. This way we don't follow sym links.
-            if (self.deleteFile(sub_path)) {
-                return;
-            } else |err| switch (err) {
-                error.FileNotFound => return,
-                error.IsDir => {},
-                error.AccessDenied => got_access_denied = true,
-
-                error.InvalidUtf8,
-                error.SymLinkLoop,
-                error.NameTooLong,
-                error.SystemResources,
-                error.ReadOnlyFileSystem,
-                error.NotDir,
-                error.FileSystem,
-                error.FileBusy,
-                error.BadPathName,
-                error.Unexpected,
-                => |e| return e,
+        const StackItem = struct {
+            name: []const u8,
+            parent_dir: Dir,
+            iter: IterableDir.Iterator,
+        };
+
+        var stack = std.BoundedArray(StackItem, 16){};
```

Comment:
The 16 here is totally arbitrary, no idea what types of things should be taken into account when determining what this size should be.

Might be worth noting that `@sizeOf(StackItem)` is `8248` on Linux (all platforms will have similar sizes because of the `buf` field of `IterableDir.Iterator`), so `@sizeOf(std.BoundedArray(StackItem, 16))` ends up being `131976` or ~128.9KiB

Reply:
It seems to me that 8 KiB is quite large for the buffer used by `IterableDir.Iterator`. Perhaps that should be reduced?

Reply:
What do you think a reasonable limit would be here? A BoundedArray size of `4` would be ~32KiB for reference.

Reply:
>  EDIT#2: Seems to be to allow for reading a lot of dirent's per getdents64 (or the equivalent) syscall.

Yes, that seems to be the idea, though my sense is that 1-2 KiB would be pretty much as effective as the current 8 in practice.

Reply:
> my sense is that 1-2 KiB would be pretty much as effective as the current 8 in practice.

You appear to be correct. `1024` and `2048` are pretty much the same from my minimal testing on Linux, and `8196` only wins out on directories that have huge numbers of entries (e.g. 25,000 files in a single directory):

Benchmarks on the dir I've been using (41209 files from `npx create-react-app my-app`):
```
Benchmark #1: ./walk-1024 my-app
  Time (mean ± σ):      29.5 ms ±   2.1 ms    [User: 4.5 ms, System: 24.8 ms]
  Range (min … max):    24.6 ms …  36.6 ms    94 runs
 
Benchmark #2: ./walk-2048 my-app
  Time (mean ± σ):      29.8 ms ±   1.9 ms    [User: 4.9 ms, System: 24.7 ms]
  Range (min … max):    24.9 ms …  34.5 ms    98 runs
 
Benchmark #3: ./walk-4096 my-app
  Time (mean ± σ):      30.2 ms ±   2.5 ms    [User: 4.8 ms, System: 25.2 ms]
  Range (min … max):    25.7 ms …  37.4 ms    107 runs
 
Benchmark #4: ./walk-8192 my-app
  Time (mean ± σ):      30.6 ms ±   2.6 ms    [User: 5.7 ms, System: 24.6 ms]
  Range (min … max):    26.7 ms …  41.1 ms    107 runs
 
Summary
  './walk-1024 my-app' ran
    1.01 ± 0.10 times faster than './walk-2048 my-app'
    1.03 ± 0.11 times faster than './walk-4096 my-app'
    1.04 ± 0.11 times faster than './walk-8192 my-app'
```

Benchmark on a single dir that contains 25,000 files in it:
```
Benchmark #1: ./walk-1024 dir-with-tons-of-entries
  Time (mean ± σ):       9.1 ms ±   2.0 ms    [User: 0.8 ms, System: 8.3 ms]
  Range (min … max):     5.5 ms …  13.2 ms    270 runs
 
Benchmark #2: ./walk-2048 dir-with-tons-of-entries
  Time (mean ± σ):       9.2 ms ±   2.0 ms    [User: 0.8 ms, System: 8.4 ms]
  Range (min … max):     5.4 ms …  12.5 ms    475 runs
 
Benchmark #3: ./walk-4096 dir-with-tons-of-entries
  Time (mean ± σ):       9.5 ms ±   1.8 ms    [User: 0.7 ms, System: 8.9 ms]
  Range (min … max):     5.4 ms …  12.7 ms    255 runs
 
Benchmark #4: ./walk-8192 dir-with-tons-of-entries
  Time (mean ± σ):       8.9 ms ±   2.1 ms    [User: 1.0 ms, System: 7.9 ms]
  Range (min … max):     5.3 ms …  13.0 ms    253 runs
 
Summary
  './walk-8192 dir-with-tons-of-entries' ran
    1.02 ± 0.34 times faster than './walk-1024 dir-with-tons-of-entries'
    1.04 ± 0.33 times faster than './walk-2048 dir-with-tons-of-entries'
    1.08 ± 0.33 times faster than './walk-4096 dir-with-tons-of-entries'
```

Reply:
I think 16.9 KiB is quite reasonable as a default. Thanks for taking the time to benchmark this!